Lib/fontTools/varLib/varStore.py

```diff
@@ -570,89 +568,63 @@ def VarStore_optimize(self, use_NO_VARIATION_INDEX=True, quantization=1):
             encodings.add_row(row)
             front_mapping[(major << 16) + minor] = row
 
-    # Separate encodings that have no gain (are decided) and those having
-    # possible gain (possibly to be merged into others.)
-    encodings = sorted(encodings.values(), key=_Encoding.__len__, reverse=True)
-    done_by_width = defaultdict(list)
-    todo = []
-    for encoding in encodings:
-        if not encoding.gain:
-            done_by_width[encoding.width].append(encoding)
-        else:
-            todo.append(encoding)
+    # Prepare for the main algorithm.
+    todo = sorted(encodings.values(), key=_Encoding.gain_sort_key)
+    encodings = []
 
-    # For each encoding that is possibly to be merged, find the best match
-    # in the decided encodings, and record that.
-    todo.sort(key=_Encoding.gain_sort_key, reverse=True)
-    for encoding in todo:
-        encoding._find_yourself_best_new_encoding(done_by_width)
-
-    # Walk through todo encodings, for each, see if merging it with
-    # another todo encoding gains more than each of them merging with
-    # their best decided encoding. If yes, merge them and add resulting
-    # encoding back to todo queue.  If not, move the enconding to decided
-    # list.  Repeat till done.
-    while todo:
-        encoding = todo.pop()
-        best_idx = None
-        best_gain = 0
-        for i, other_encoding in enumerate(todo):
-            combined_chars = other_encoding.chars | encoding.chars
-            combined_width = _Encoding._popcount(combined_chars)
-            combined_overhead = _Encoding._characteristic_overhead(combined_chars)
-            combined_gain = (
-                +encoding.overhead
-                + other_encoding.overhead
-                - combined_overhead
-                - (combined_width - encoding.width) * len(encoding)
-                - (combined_width - other_encoding.width) * len(other_encoding)
-            )
-            this_gain = (
-                0
-                if encoding.best_new_encoding is None
-                else (
-                    +encoding.overhead
-                    - (encoding.best_new_encoding.width - encoding.width)
-                    * len(encoding)
-                )
-            )
-            other_gain = (
-                0
-                if other_encoding.best_new_encoding is None
-                else (
-                    +other_encoding.overhead
-                    - (other_encoding.best_new_encoding.width - other_encoding.width)
-                    * len(other_encoding)
-                )
-            )
-            separate_gain = this_gain + other_gain
+    # Repeatedly pick two best encodings to combine, and combine them.
 
-            if combined_gain - separate_gain > best_gain:
-                best_idx = i
-                best_gain = combined_gain - separate_gain
+    heap = []
+    for i, other_encoding in enumerate(todo):
```

Comment:
i'm confused here, are we ever using `other_encoding`? it looks like we rebind it in the next loop without having touched it. Maybe it was supposed to be `encoding`, and then you wouldn't need to manually bind that, below?

Reply:
You are correct.